### PR TITLE
Set up associations

### PR DESCRIPTION
**Pull request by tomasesquivelgc**
# In this branch
 - New classes added (book, classroom, rental)
 - Updated relationships between classes
 - Students belong in classrooms, classrooms have students
 - Person superclass can now make rental of books

### CODE REVIEW

File: project/classroom.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+  def initialize(label)
+    @label = label
+    @students = []
+  end
+
+  def add_student(student)
+    @students << student
+    student.classroom = self
+  end
```
Comment (danmainah) [inline project/classroom.rb]: - [x] @tomasesquivelgc Great work adding the classroom class and adding the has-many side (a classroom has many students) of the relationship between Classroom and Student, However, you forgot to add the Getter for students, kindly consider adding it.

File: Project/student.rb [hunk 2]
```diff
@@ -9,4 +12,9 @@ def initialize(age, classroom, name = 'Unknown', parent_permission: true)
   def play_hooky
     '¯(ツ)/¯'
   end
+
+  def classroom=(classroom)
+    @classroom = classroom
+    classroom.add_student(self)
```
Comment (danmainah) [inline Project/student.rb]: - [x] @tomasesquivelgc Great work adding the relationship between the belongs-to side (a student belongs to a classroom)of the relationship between Classroom and Student, however, I would suggest a slight change in your code to include the situation when the student is already in the class and there is no need to add the student.
``` suggestion
    classroom.students.push(self) unless classroom.students.include?(self)
```

File: project/book.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rental = []
+  end
+
+  def add_rental(rental)
+    @rental << rental
+    rental.book = self
+  end
```
Comment (danmainah) [inline project/book.rb]: - [x] @tomasesquivelgc Great work adding a method to implement the many-to-many relationships between `Person` and `Book` through rentals, however we need to also pass the date and person as parameters on top of the book itself when creating a rental
``` suggestion
  def add_rental(date, person)
    Rental.new(date, self, person)
  end
```

File: project/rental.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class Rental
+  attr_accessor :date
+  attr_reader :book, :person
```
Comment (danmainah) [inline project/rental.rb]: - [x] @tomasesquivelgc kindly move your book and person variable from  attr_reader to  attr_accessordince we are not only reading them but also modifying them.

File: Project/person.rb [hunk 5]
```diff
@@ -21,6 +22,11 @@ def correct_name
     @name
   end
 
+  def add_rental(rental)
+    @rental << rental
+    rental.person = self
+  end
```
Comment (danmainah) [inline Project/person.rb:28]: - [x] @tomasesquivelgc Great work adding a method to implement the many-to-many relationships between Person and Book through rentals, however we need to also pass the date and book as parameters on top of the person when creating a rental
``` suggestion
  def add_rental(date, book)
     Rental.new(date, book, self)
  end
```

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @tomasesquivelgc , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️ Created book and rental classes and ensured the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental is implemented.
✔️ Created classroom class and ensured the has-many side (a classroom has many students) relationship has been implemented.
✔️ Added new method in student class and ensured the belongs-to side (a classroom has many students) relationship has been implemented.

# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Review by ichala — APPROVED** [review]
Hi Student  ,
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:  
 
68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f6c3366516333614356304871707a5763552f67697068792e676966

# To Highlight 💪🏻

- Changes Implemented ✔️ 


 Good job implementing the requirements  :+1: Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Optional Suggestions  :

- N/A


**********
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag  ichala in your question so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**